### ISSUE TITLE
Is multiplication by negmask in one_hot2dist() irrelevant?

### ISSUE THREAD

**Issue by dtoniolo** [LIVIAETS/boundary-loss#49]
First of all, thank you for your work and for making the code publicly available. I find it very interesting and I have a small question about the [implementation of `one_hot2dist()`](https://github.com/LIVIAETS/boundary-loss/blob/749602d822fcb097c79a7fb708805b8982c37030/utils.py#L271):
```python
def one_hot2dist(seg: np.ndarray, resolution: Tuple[float, float, float] = None,
                 dtype=None) -> np.ndarray:
    assert one_hot(torch.tensor(seg), axis=0)
    K: int = len(seg)

    res = np.zeros_like(seg, dtype=dtype)
    for k in range(K):
        posmask = seg[k].astype(np.bool)

        if posmask.any():
            negmask = ~posmask
            res[k] = eucl_distance(negmask, sampling=resolution) * negmask \
                - (eucl_distance(posmask, sampling=resolution) - 1) * posmask
        # The idea is to leave blank the negative classes
        # since this is one-hot encoded, another class will supervise that pixel

    return res
```
In the following, for the sake of simplicity I'll define `euc_neg = eucl_distance(negmask, sampling=resolution)` and `euc_pos = eucl_distance(posmask, sampling=resolution)`.

The result of `eucl_distance(negmask, sampling=resolution)` is multiplied by `negmask`, but this seems to be not necessary: `negmask` is zero when `euc_neg` is zero, and is one otherwise. So `euc_neg` should be unchanged. The multiplication instead seems to be needed for the other term, in order to remove the `-1` values on the pixels that belong to the k-th class.

Am I getting something wrong or could the multiplication by `negmask` be removed without consequence?

The multiplication by `negmask` admittedly isn't a very computationally intensive operation, so removing it wouldn't change much, I'm more concerned about making sure I understood the loss definition and implementation correctly.

Thanks again for the great work.

**Comment by HKervadec** [comment]
Hey,

thanks for the interest. 

I haven't played with it for some time, and I wrote this three years ago, so I am not fully sure anymore of the rationale behind this. But I'll do my best.

> Am I getting something wrong or could the multiplication by negmask be removed without consequence?

This seems mostly correct, my rationale at the time was (I think):
- I did not fully trust this assumption was right for _every_ corner case ; 
- I was not even sure that I would keep using the scipy distance transform ;
- I think it makes the code slightly more readable, by highlighting that it is two independent distance computations that are then combined.

So in the end, adding this extra (probably redundant) multiplication adds some trust in the code (at least for me [1]), and makes the code more "symmetric" (both the positive and negative parts are handled in the same way).

Best,

Hoel

---
[1] This is especially true when you come back at it after a few months or even years ; I can check its content without wondering too much about the exact details of the scipy distance transform.
